From 5c584d98c1ba473ecba08930d0c278033bef69b5 Mon Sep 17 00:00:00 2001
From: Bai Ping <b51503@freescale.com>
Date: Sat, 28 Feb 2015 19:19:56 +0800
Subject: [PATCH 0477/5242] MLK-10271 cpufreq: imx6: improve busfreq operation
 when wehn setpoint lower than 396MHz

commit  81e84172ec6f488fab1a07b7d51098b7897c8819 from
https://source.codeaurora.org/external/imx/linux-imx.git

for i.MX6SX, according to the latest datasheet, added a 198MHz setpoint in cpufreq driver.
The 198MHz setpoint is NOT enough to support playing mp3,the system will stay at a higher
setpoint and high_bus_mode. So when having a setpoint lower than 396MHz, make sure when
the cpufreq is at 396MHz or lower, the busfreq is always in low_bus_mode to save more power.

Signed-off-by: Bai Ping <b51503@freescale.com>
(cherry picked from commit 3ba9548200ffb2c85111dd84946046ae0c7b09c4)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/cpufreq/imx6q-cpufreq.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/cpufreq/imx6q-cpufreq.c b/drivers/cpufreq/imx6q-cpufreq.c
index 231109d..8c3d29a 100644
--- a/drivers/cpufreq/imx6q-cpufreq.c
+++ b/drivers/cpufreq/imx6q-cpufreq.c
@@ -21,6 +21,7 @@
 #define PU_SOC_VOLTAGE_NORMAL	1250000
 #define PU_SOC_VOLTAGE_HIGH	1275000
 #define FREQ_1P2_GHZ		1200000000
+#define FREQ_396_MHZ		396000
 
 static struct regulator *arm_reg;
 static struct regulator *pu_reg;
@@ -95,7 +96,7 @@ static int imx6q_set_target(struct cpufreq_policy *policy, unsigned int index)
 	 * CPU freq is increasing, so need to ensure
 	 * that bus frequency is increased too.
 	 */
-	if (old_freq == freq_table[0].frequency)
+	if (old_freq <= FREQ_396_MHZ && new_freq > FREQ_396_MHZ)
 		request_bus_freq(BUS_FREQ_HIGH);
 
 	/* scaling up?  scale voltage before frequency */
@@ -216,7 +217,7 @@ static int imx6q_set_target(struct cpufreq_policy *policy, unsigned int index)
 	 * If CPU is dropped to the lowest level, release the need
 	 * for a high bus frequency.
 	 */
-	if (new_freq == freq_table[0].frequency)
+	if (old_freq > FREQ_396_MHZ && new_freq <= FREQ_396_MHZ)
 		release_bus_freq(BUS_FREQ_HIGH);
 
 	return 0;
@@ -234,9 +235,8 @@ static int imx6q_cpufreq_init(struct cpufreq_policy *policy)
 		dev_err(cpu_dev, "imx6 cpufreq init failed!\n");
 		return ret;
 	}
-	if (policy->cur > freq_table[0].frequency)
+	if (policy->cur > FREQ_396_MHZ)
 		request_bus_freq(BUS_FREQ_HIGH);
-
 	return 0;
 }
 
-- 
1.7.9.5

